### PR TITLE
RowViewListModel

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

**Review by spireken — APPROVED** [review]

File: Applications/Spire/Include/Spire/Ui/RowViewListModel.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      const std::any& at(int index) const override;
+
+      QValidator::State set(int index, const std::any& value) override;
+
+      boost::signals2::connection connect_operation_signal(
+        const typename OperationSignal::slot_type& slot) const override;
+
+    private:
+      std::shared_ptr<TableModel> m_source;
+      boost::optional<int> m_row;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/RowViewListModel.hpp:37]: Use a sentinel value of ```-1``` for ```m_row```.

File: Applications/Spire/Source/Ui/RowViewListModel.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+
+RowViewListModel::RowViewListModel(std::shared_ptr<TableModel> source, int row)
+    : m_source(std::move(source)),
+      m_row(row) {
+  if(row < 0 || row >= m_source->get_row_size()) {
+    m_row = none;
```
Comment (kranar) [inline Applications/Spire/Source/Ui/RowViewListModel.cpp:11]: When ```m_row == -1```, do the following:

    m_source_connection.disconnect();
    m_source = nullptr;

File: Applications/Spire/Source/Ui/RowViewListModel.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  }
+  m_source_connection = m_source->connect_operation_signal(
+    [=] (const auto& operation) { on_operation(operation); });
+}
+
+int RowViewListModel::get_size() const {
+  if(!m_row) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/RowViewListModel.cpp:18]: ```m_row == -1```

File: Applications/Spire/Source/Ui/RowViewListModel.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+connection RowViewListModel::connect_operation_signal(
+    const typename OperationSignal::slot_type& slot) const {
+  return m_transaction.connect_operation_signal(slot);
+}
+
+void RowViewListModel::on_operation(const TableModel::Operation& operation) {
+  m_transaction.transact([&] {
+    visit<TableModel>(operation,
+      [&] (const TableModel::AddOperation& operation) {
+        if(m_row && *m_row >= operation.m_index) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/RowViewListModel.cpp:47]: You can remove all of the ```if(m_row ...)``` checks. For ```m_row == -1``` none of the remainder of the condition will hold.

**Review by kranar — CHANGES_REQUESTED** [review]

**Review by kranar — APPROVED** [review]
Excellent stuff, well done.

**Pull request merged**